### PR TITLE
Add function for retrieving token recipient senders

### PR DESCRIPTION
**Pull request by dylanjw**
ref: https://github.com/gitcoinco/web/issues/6395

##### Description
Add a function `get_token_recipient_senders` to `dashboard.utils`. It will retrieve and filter event logs for erc20 token transfers. Always returns a list.

##### Refers/Fixes

https://github.com/gitcoinco/web/issues/6395

##### Testing

Let me know if you want tests written. I was unsure how to deal with the set up of test blockchain state. Here is a gist with a few assertions against some addresses I set up on rinkeby: https://gist.github.com/dylanjw/f3c0d876aee170b71a61858a322d24ed

### CODE REVIEW

**Comment by danlipert** [comment]
@dylanjw thanks for the PR! Generally, for testing, we either would like to see unit tests (mocking API responses usually works fine) or just describe how you tested or show your test cases like with your gist 👍

File: app/dashboard/utils.py [hunk 1]
```diff
@@ -998,3 +999,33 @@ def list_urls(lis, acc=None):
...
+    if address == '0x0000000000000000000000000000000000000000':
+        return '0x0'
+    else:
+        return address
+
+
+def get_token_recipient_senders(recipient_address, token_addres):
+    w3 = Web3(HTTPProvider(settings.WEB3_HTTP_PROVIDER))
+    contract = w3.eth.contract(
+        address=token_address,
```
Comment (owocki) [inline app/dashboard/utils.py:1014]: token_address is not defined

File: app/dashboard/utils.py [hunk 2]
```diff
@@ -998,3 +999,33 @@ def list_urls(lis, acc=None):
...
+        abi=erc20_abi,
+    )
+
+    balance = contract.functions.balanceOf(recipient_address).call()
+
+    if balance == 0:
+        return []
+
+    transfers = contract.events.Transfer.getLogs(
```
Comment (owocki) [inline app/dashboard/utils.py:1023]: AttributeError: type object 'Transfer' has no attribute 'getLogs'

Comment (owocki) [reply app/dashboard/utils.py:1023]: what version of web3py are you running? im running 4.6.0, which is the version in the requirements/base.txt

Comment (dylanjw) [reply app/dashboard/utils.py:1023]: Oops, I wrote this using the latest web3. Ill get this fixed later today.

File: app/dashboard/utils.py [hunk 3]
```diff
@@ -998,3 +1000,68 @@ def list_urls(lis, acc=None):
...
+
+    contract = w3.eth.contract(
+        address=token_address,
+        abi=erc20_abi,
+    )
+
+    balance = contract.functions.balanceOf(recipient_address).call()
+
+    if balance == 0:
+        return []
```
Comment (dylanjw) [inline app/dashboard/utils.py:1048]: I probably should take this out. If a user receives tokens, and then transfers them out; this function will return `[]` as if they never received them. I intended to create a fast exit path, but to determine if a address has ever received tokens, I dont think I can avoid retrieving the token transfer event logs.

**Comment by dylanjw** [comment]
@owocki Ive updated the function to work with web3, 4.6, and added a test. Let me know if you need any other changes.

**Comment by dylanjw** [comment]
@owocki Are you planning to upgrade to web3 >=5 soon? I have time to help if that is on the todo list.

**Comment by owocki** [comment]
defer to @danlipert and @gitcoinco/engineers on that

**Comment by thelostone-mc** [comment]
@dylanjw https://github.com/gitcoinco/web/pull/6366

**Pull request merged**